src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class AbstractFlowPass<TLocalState, TLocalFunctionState>
+    {
+        protected sealed class PendingBranchesCollection : IEnumerable<PendingBranch>
+        {
+            private ArrayBuilder<PendingBranch> _branches;
+            private PooledDictionary<LabelSymbol, ArrayBuilder<PendingBranch>>? _branchesToLabel;
+
+            internal PendingBranchesCollection()
+            {
+                _branches = ArrayBuilder<PendingBranch>.GetInstance();
+            }
+
+            internal void Free()
+            {
+                _branches.Free();
+                _branches = null!;
+                FreeBranchesToLabel();
+            }
+
+            internal void Clear()
+            {
+                _branches.Clear();
+                FreeBranchesToLabel();
+            }
+
+            private void FreeBranchesToLabel()
+            {
+                if (_branchesToLabel is { })
+                {
+                    foreach (var branches in _branchesToLabel.Values)
+                    {
+                        branches.Free();
+                    }
+                    _branchesToLabel.Free();
+                    _branchesToLabel = null;
+                }
+            }
+
+            internal ImmutableArray<PendingBranch> ToImmutable()
+            {
+                return _branchesToLabel is null ?
+                    _branches.ToImmutable() :
+                    ImmutableArray.CreateRange(this);
+            }
+
+            internal ArrayBuilder<PendingBranch> GetAndRemoveBranches(LabelSymbol? label)
+            {
+                ArrayBuilder<PendingBranch>? result;
+                if (label is null)
+                {
+                    result = _branches;
+                    _branches = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                else if (_branchesToLabel is { } && _branchesToLabel.TryGetValue(label, out result))
+                {
+                    _branchesToLabel.Remove(label);
+                }
+                else
+                {
+                    result = ArrayBuilder<PendingBranch>.GetInstance();
```

Comment:
>result = ArrayBuilder<PendingBranch>.GetInstance(); [](http://example.com/codeflow?start=20&length=51)

It feels like we can easy avoid allocating empty builders here and above and return null instead.

Reply:
It feels like we can also do the same above, when ```label is null``` and ```_unlabeledBranches``` is empty.